apps/aecore/lib/aecore/peers/sync.ex

```diff
@@ -98,11 +159,102 @@ defmodule Aecore.Peers.Sync do
     |> Enum.shuffle
     |> Enum.take(Enum.min([@peers_target_count - known_count, known_count]))
     |> Enum.reduce(0, fn(peer, acc) ->
-      case Peers.add_peer(peer) do
+      peer_uri = elem(peer, 0)
+      case Peers.add_peer(peer_uri) do
         :ok -> acc+1
         _ -> acc
       end
     end)
   end
-end
 
+  defp build_chain(state, {block_hash, peer}, chain) do
+    case(HttpClient.get_block({peer, Base.encode16(block_hash)})) do
+      {:ok, peer_block} ->
+        deserialized_block = Serialization.block(peer_block, :deserialize)
+        has_parent_block_in_state = Map.has_key?(state, deserialized_block.header.prev_hash)
+        has_parent_in_chain =
+          deserialized_block.header.prev_hash == BlockValidation.block_header_hash(Chain.latest_block().header)
+        cond do
+          has_parent_block_in_state ->
+            build_chain(state, {deserialized_block.header.prev_hash, peer}, [deserialized_block | chain])
+          has_parent_in_chain ->
+            [deserialized_block | chain]
+          true ->
+            []
+        end
+      :error ->
+        Logger.info(fn -> "Couldn't get block #{block_hash} from #{peer}" end)
+    end
+  end
+
+  defp add_built_chain(chain) do
+    Enum.each(chain, fn(block) ->
+        Chain.add_block(block)
+      end)
+  end
+
+  defp remove_added_blocks_from_state(state) do
+    state_with_removed_blocks = Enum.filter(state, fn {block_hash, _} ->
+        case Chain.get_block(block_hash) do
+          {:error, _} ->
+            true
+          _ ->
+            false
+        end
+      end)
+
+    List.foldl(state_with_removed_blocks, %{}, fn({block_hash, block_data}, acc) ->
+        Map.put(acc, block_hash, block_data)
+      end)
+  end
+
+  defp check_peer_block(peer_uri, block_hash, blocks_with_status) do
+    case Chain.get_block_by_hex_hash(block_hash) do
+      {:error, _} ->
+        case(HttpClient.get_block({peer_uri, block_hash})) do
+          {:ok, peer_block} ->
+            deserialized_block = Serialization.block(peer_block, :deserialize)
+            peer_block_hash =
+              BlockValidation.block_header_hash(deserialized_block.header)
+            status =
+              case(HttpClient.get_block({peer_uri, Base.encode16(peer_block_hash)})) do
+                {:ok, _peer_block_parent} ->
+                  :good
+                :error ->
+                  :bad
+              end
+
```

Comment:
We should *single_validate* the blocks here. This way:
- we don't store info about invalid blocks
- we *single_validate* each block only ones.